packages/eventual-send/src/index.js

```diff
@@ -71,33 +68,109 @@ export function makeHandledPromise() {
       while (target !== p) {
         const parent = forwardedPromiseToPromise.get(target);
         forwardedPromiseToPromise.delete(target);
-        promiseToUnsettledHandler.delete(target);
+        promiseToPendingHandler.delete(target);
         promiseToPresence.set(target, presence);
         target = parent;
       }
     } else {
-      // We propagate p and remove all other unsettled handlers
+      // We propagate p and remove all other pending handlers
       // upstream.
       // Note that everything except presences is covered here.
       while (target !== p) {
         const parent = forwardedPromiseToPromise.get(target);
         forwardedPromiseToPromise.set(target, p);
-        promiseToUnsettledHandler.delete(target);
+        promiseToPendingHandler.delete(target);
         target = parent;
       }
     }
     return target;
   }
 
-  // This special handler accepts Promises, and forwards
-  // handled Promises to their corresponding fulfilledHandler.
+  /**
+   * This special handler accepts Promises, and forwards
+   * handled Promises to their corresponding fulfilledHandler.
+   *
+   * @type {Required<EHandler<any>>}
+   */
   let forwardingHandler;
   let handle;
 
-  function HandledPromise(executor, unsettledHandler = undefined) {
-    if (new.target === undefined) {
-      throw new Error('must be invoked with "new"');
+  /**
+   * @param {string} handlerName
+   * @param {EHandler<any>} handler
+   * @param {string} operation
+   * @param {any} o
+   * @param {any[]} opArgs
+   * @param {Promise<unknown>} [returnedP]
+   * @returns {any}
+   */
+  const dispatchToHandler = (
+    handlerName,
+    handler,
+    operation,
+    o,
+    opArgs,
+    returnedP,
+  ) => {
+    let actualOp = operation;
+    const triedOperations = [operation];
+    const isSendOnly = actualOp.endsWith('SendOnly');
+    const makeResult = result => (isSendOnly ? undefined : result);
+
+    if (isSendOnly) {
+      // We don't specify the resulting promise if it is sendonly.
+      returnedP = undefined;
+    }
+
+    if (isSendOnly && typeof handler[actualOp] !== 'function') {
+      // Substitute for sendonly with the corresponding non-sendonly operation.
+      actualOp = actualOp.slice(0, -'SendOnly'.length);
+      triedOperations.push(actualOp);
+    }
+
+    // Fast path: just call the actual operation.
+    if (typeof handler[actualOp] === 'function') {
+      return makeResult(handler[actualOp](o, ...opArgs, returnedP));
     }
+
+    if (actualOp === 'applyMethod') {
+      // Compose a missing applyMethod by get followed by applyFunction.
+      const [prop, args] = opArgs;
+      const getResultP = handle(o, 'get', [prop], undefined);
```

Comment:
Given the code below that downlevels an `applyFunction` to an `applyMethod` with a property name of `undefined`, our `applyMethod` logic must be compat with a property name of `undefined`. Here, we downlevel  it further to a `get` followed by an `applyFunction`. Is this `get` compat with `prop === undefined`?

Reply:
You pointed out this case where another `coerceToObjectProperty` is needed before calling `get`.  Added.

Reply:
How does this fix the recursion? What if both `applyFunction` and `applyMethod` are absent and `actualOp === 'applyFunction'`? AFAICT, nothing catches that case early. The second case fires first. As you say, it
```js
Downlevel a missing applyFunction to applyMethod with undefined name
```
But in the `applyFunction` base case, `coerceToObjectProperty(undefined) === 'undefined'`, so it looks to me like it continues with the property name `'undefined'` rather than `undefined`. It bottoms out the recursion where it should, but with the wrong operation.